### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Skeletal 

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #2394

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2394
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2394~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Skeletal.lean [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+variable (D : Type u₂) [Category.{v₂} D]
+
+variable {E : Type u₃} [Category.{v₃} E]
+
+/-- A category is skeletal if isomorphic objects are equal. -/
+def Skeletal : Prop :=
+  ∀ ⦃X Y : C⦄, IsIsomorphic X Y → X = Y
+#align category_theory.skeletal CategoryTheory.Skeletal
+
+/-- `is_skeleton_of C D F` says that `F : D ⥤ C` exhibits `D` as a skeletal full subcategory of `C`,
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Skeletal.lean:49]:
```suggestion
/-- `IsSkeletonOf C D F` says that `F : D ⥤ C` exhibits `D` as a skeletal full subcategory of `C`,
```

File: Mathlib/CategoryTheory/Skeletal.lean [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+variable {C D}
+
+/-- If `C` is thin and skeletal, then any naturally isomorphic functors to `C` are equal. -/
+theorem Functor.eq_of_iso {F₁ F₂ : D ⥤ C} [Quiver.IsThin C] (hC : Skeletal C) (hF : F₁ ≅ F₂) :
+    F₁ = F₂ :=
+  Functor.ext (fun X => hC ⟨hF.app X⟩) fun _ _ _ => Subsingleton.elim _ _
+#align category_theory.functor.eq_of_iso CategoryTheory.Functor.eq_of_iso
+
+/-- If `C` is thin and skeletal, `D ⥤ C` is skeletal.
+`category_theory.functor_thin` shows it is thin also.
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Skeletal.lean:70]:
```suggestion
`CategoryTheory.functor_thin` shows it is thin also.
```

File: Mathlib/CategoryTheory/Skeletal.lean [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+`map₂` by currying.
+-/
+/-- Given a bifunctor, we descend to a function on objects of `ThinSkeleton` -/
+def map₂ObjMap (F : C ⥤ D ⥤ E) : ThinSkeleton C → ThinSkeleton D → ThinSkeleton E := 
+  fun x y =>
+    @Quotient.map₂ C D (isIsomorphicSetoid C) (isIsomorphicSetoid D) E (isIsomorphicSetoid E) 
+      (fun X Y => (F.obj X).obj Y)
+          (fun X₁ _ ⟨hX⟩ _ Y₂ ⟨hY⟩ => ⟨(F.obj X₁).mapIso hY ≪≫ (F.mapIso hX).app Y₂⟩) x y
+
+/-- For each `x : ThinkSkeleton C`, we promote `map₂ObjMap F x` to a functor -/
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Skeletal.lean:229]:
```suggestion
/-- For each `x : ThinSkeleton C`, we promote `map₂ObjMap F x` to a functor -/
```

File: Mathlib/CategoryTheory/Skeletal.lean [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+/-- This provides natural transformations `map₂Functor F x₁ ⟶  map₂Functor F x₂` given 
+`x₁ ⟶  x₂` -/
+def map₂NatTrans (F : C ⥤ D ⥤ E) : {x₁ x₂ : ThinSkeleton C} → (x₁ ⟶  x₂) →  
+    (map₂Functor F x₁ ⟶  map₂Functor F x₂) := fun {x₁} {x₂} => 
+  @Quotient.recOnSubsingleton₂ C C (isIsomorphicSetoid C) (isIsomorphicSetoid C) 
+    (fun x x' : ThinSkeleton C => (x ⟶  x') → (map₂Functor F x ⟶  map₂Functor F x')) _ x₁ x₂ 
+    (fun X₁ X₂ f => { app := fun y =>   
+      Quotient.recOnSubsingleton y fun Y => homOfLE (f.le.elim fun f' => ⟨(F.map f').app Y⟩) }) 
+
+-- TODO: state the lemmas about what happens when you compose with `to_thin_skeleton`
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Skeletal.lean:247]:
```suggestion
-- TODO: state the lemmas about what happens when you compose with `toThinSkeleton`
```

**Review by jcommelin — COMMENTED** [review]
Please fix other docstrings too...

File: Mathlib/CategoryTheory/Skeletal.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Skeletal.lean]: Done

File: Mathlib/CategoryTheory/Skeletal.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Skeletal.lean]: Done

File: Mathlib/CategoryTheory/Skeletal.lean (hunk 3 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Skeletal.lean:229]: Done

File: Mathlib/CategoryTheory/Skeletal.lean (hunk 4 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Skeletal.lean]: Done

**Comment by mattrobball** [comment]
Fixed the docstrings. Should be good now.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**